### ISSUE TITLE
GC segfault (no MWE, private repo)

### ISSUE THREAD

**Issue by tpapp** [EnzymeAD/Enzyme.jl#853]
I am getting a GC segfault (Enzyme#main, julia 1.9) on a private repo (that was used to produce #817, #840, #841). Individual calls to `autodiff` are fine now, but benchmarking with BenchmarkTools or MCMC sampling with DynamicHMC produces non-deterministic segfaults. I can provide the commit reference in a repo (necessarily large, not an MWE) or try to reduce this down to an MWE, but I am not sure who feasible that is.

I get these warnings too, not sure how relevant:
```julia
┌ Warning: Type does not have a definite number of fields
│   T = Tuple
└ @ Enzyme ~/.julia/packages/GPUCompiler/BxfIW/src/utils.jl:56
warning: didn't implement memmove, using memcpy as fallback which can result in errors
warning: didn't implement memmove, using memcpy as fallback which can result in errors
┌ Warning: Type does not have a definite number of fields
│   T = Tuple
└ @ Enzyme ~/.julia/packages/GPUCompiler/BxfIW/src/utils.jl:56
warning: didn't implement memmove, using memcpy as fallback which can result in errors
warning: didn't implement memmove, using memcpy as fallback which can result in errors
```

This is the segfault info:

```julia
[3126961] signal (11.128): Segmentation fault
in expression starting at /home/tamas/code/julia/misc/EnzymeBugMWE.jl/replicate_crash.jl:123
gc_try_setmark at /home/tamas/src/julia-1.9/src/gc.c:1965 [inlined]
gc_mark_scan_obj16 at /home/tamas/src/julia-1.9/src/gc.c:2250 [inlined]
gc_mark_loop at /home/tamas/src/julia-1.9/src/gc.c:2526
_jl_gc_collect at /home/tamas/src/julia-1.9/src/gc.c:3400
ijl_gc_collect at /home/tamas/src/julia-1.9/src/gc.c:3707
maybe_collect at /home/tamas/src/julia-1.9/src/gc.c:1078 [inlined]
jl_gc_pool_alloc_inner at /home/tamas/src/julia-1.9/src/gc.c:1443 [inlined]
jl_gc_pool_alloc_noinline at /home/tamas/src/julia-1.9/src/gc.c:1504 [inlined]
jl_gc_alloc_ at /home/tamas/src/julia-1.9/src/julia_internal.h:460 [inlined]
jl_gc_alloc at /home/tamas/src/julia-1.9/src/gc.c:3754
unknown function (ip: 0x7fc487831d1b)
unknown function (ip: 0x7fc487835142)
Allocations: 3249686358 (Pool: 3226128646; Big: 23557712); GC: 11152
```

**Comment by tpapp** [comment]
A variant of the error message is
```julia
GC error (probable corruption) :
Allocations: 75254965 (Pool: 75128539; Big: 126426); GC: 111

!!! ERROR in jl_ -- ABORTING !!!
0x7f3e316ba010: Queued root: 0x7f3e2e2ea8b0 :: 0x7f3e81f92660 (bits: 3)
        of type Task
0x7f3e316ba028: Queued root: 0x7f3e2e2eabd0 :: 0x7f3e81f92660 (bits: 3)
        of type Task
0x7f3e316ba040: Queued root: 0x7f3e2e2eaa40 :: 0x7f3e81f92660 (bits: 3)
        of type Task
0x7f3e316ba058: Queued root: 0x7f3e8692c010 :: 0x7f3e81f92660 (bits: 3)
        of type Task
0x7f3e316ba070: Queued root: 0x7f3e86924010 :: 0x7f3e81f92660 (bits: 3)
        of type Task
0x7f3e316ba088: Queued root: 0x7f3e86920010 :: 0x7f3e81f92660 (bits: 3)
        of type Task
0x7f3e316ba0a0: Queued root: 0x7f3e86918010 :: 0x7f3e81f92660 (bits: 3)
        of type Task
0x7f3e316ba0b8: Queued root: 0x7f3e86928010 :: 0x7f3e81f92660 (bits: 3)
        of type Task
0x7f3e316ba0d0: Queued root: 0x7f3e86914010 :: 0x7f3e81f92660 (bits: 3)
        of type Task
0x7f3e316ba0e8: Queued root: 0x7f3e86824650 :: 0x7f3e81f92660 (bits: 3)
        of type Task
0x7f3e316ba100: Queued root: 0x7f3e868244c0 :: 0x7f3e81f92660 (bits: 3)
        of type Task
0x7f3e316ba118: Queued root: 0x7f3e8691c010 :: 0x7f3e81f92660 (bits: 3)
        of type Task
0x7f3e316ba130:  r-- Stack frame 0x7ffda86c7280 -- 0 of 1444 (direct)
0x7f3e316ba158:   `- Object (16bit) 0x7f3e0463ea40 :: 0x7f3e05fb0551 -- [12, 24)
        of type Tuple{NamedTuple{(Symbol("1"), Symbol("2"), Symbol("3"), Symbol("4"), Symbol("5"), Symbol("6"), Symbol("7"), Symbol("8"), Symbol("9"), Symbol("10"), Symbol("11"), Symbol("12"), Symbol("13"), Symbol("14"), Symbol("15"), Symbol("16"), Symbol("17"), Symbol("18"), Symbol("19"), Symbol("20"), Symbol("21"), Symbol("22"), Symbol("23"), Symbol("24"), Symbol("25"), Symbol("26"), Symbol("27"), Symbol("28")), Tuple{Float64, Tuple{NTuple{9, Float64}}, Tuple{NTuple{9, Float64}}, Tuple{Tuple{Float64, Float64, Float64}}, Vararg{Any, 24}}}, Float64, UInt64}

[3161021] signal (6.-6): Aborted
in expression starting at REPL[37]:1
pthread_kill at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
gc_assert_datatype_fail at /home/tamas/src/julia-1.9/src/gc.c:1912
gc_mark_loop at /home/tamas/src/julia-1.9/src/gc.c:3020
_jl_gc_collect at /home/tamas/src/julia-1.9/src/gc.c:3400
ijl_gc_collect at /home/tamas/src/julia-1.9/src/gc.c:3707
maybe_collect at /home/tamas/src/julia-1.9/src/gc.c:1078 [inlined]
jl_gc_pool_alloc_inner at /home/tamas/src/julia-1.9/src/gc.c:1443 [inlined]
jl_gc_pool_alloc_noinline at /home/tamas/src/julia-1.9/src/gc.c:1504 [inlined]
jl_gc_alloc_ at /home/tamas/src/julia-1.9/src/julia_internal.h:460 [inlined]
jl_gc_alloc at /home/tamas/src/julia-1.9/src/gc.c:3754
unknown function (ip: 0x7f3df5e2e236)
unknown function (ip: 0x7f3df5e15830)
Allocations: 75254965 (Pool: 75128539; Big: 126426); GC: 111
```

**Comment by tpapp** [comment]
Replication is in https://github.com/tpapp/EnzymeBugMWE.jl/pull/3.

**Comment by tpapp** [comment]
With latest Enzyme#main (7da242da), I am getting 

```julia
free(): invalid next size (fast)
```
which itself does not segfault but in a subsequent call

```
[322513] signal (6.-6): Aborted
in expression starting at REPL[15]:1
pthread_kill at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x7f8ffeace5a3)
unknown function (ip: 0x7f8ffeb46146)
unknown function (ip: 0x7f8ffeb4803b)
__libc_free at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x7f8fa28abfc7)
unknown function (ip: 0x7f8fa28ae1d5)
unknown function (ip: 0x7f8fa28ab9c2)
unknown function (ip: 0x7f8fa28aa774)
macro expansion at /home/tamas/.julia/packages/Enzyme/tHbH4/src/compiler.jl:9508 [inlined]
enzyme_call at /home/tamas/.julia/packages/Enzyme/tHbH4/src/compiler.jl:9200 [inlined]
CombinedAdjointThunk at /home/tamas/.julia/packages/Enzyme/tHbH4/src/compiler.jl:9163 [inlined]
autodiff at /home/tamas/.julia/packages/Enzyme/tHbH4/src/Enzyme.jl:205 [inlined]
autodiff at /home/tamas/.julia/packages/Enzyme/tHbH4/src/Enzyme.jl:214 [inlined]
logdensity_and_gradient at /home/tamas/.julia/packages/LogDensityProblemsAD/LIUcJ/ext/LogDensityProblemsADEnzymeExt.jl:73 [inlined]
#evaluate_ℓ#2 at /home/tamas/.julia/packages/DynamicHMC/pC62v/src/hamiltonian.jl:204
evaluate_ℓ at /home/tamas/.julia/packages/DynamicHMC/pC62v/src/hamiltonian.jl:202 [inlined]
#initialize_warmup_state#27 at /home/tamas/.julia/packages/DynamicHMC/pC62v/src/mcmc.jl:125 [inlined]
initialize_warmup_state at /home/tamas/.julia/packages/DynamicHMC/pC62v/src/mcmc.jl:123 [inlined]
#mcmc_keep_warmup#34 at /home/tamas/.julia/packages/DynamicHMC/pC62v/src/mcmc.jl:496
free(): invalid next size (fast)
```

does.

Not sure if I should open a separate issue or if it is the same.

**Comment by tpapp** [comment]
With recent versions of Enzyme it is still the `invalid next size (fast)` bug. A replicator is in https://github.com/tpapp/EnzymeBugMWE.jl/pull/3.

**Comment by tpapp** [comment]
https://github.com/tpapp/EnzymeBugMWE.jl/pull/5 replicates the GC segfault, again.

Since this is a GC-related bug, should I ask for help on the Julia forum debugging this?

**Comment by wsmoses** [comment]
[zlog.txt](https://github.com/EnzymeAD/Enzyme.jl/files/11560877/zlog.txt)

**Comment by wsmoses** [comment]
@vchuravy and I managed to debug the root cause of this to be a generic bug in LLVM itself wrt tail calls. Aka not Julia, Enzyme, or anything else but a bug that could impacat all programming languages/etc again independent of AD.

In the jll bump today we added some code that seems to remedy it here, but still we're uneasy about the internal LLVM miscompile without the workaround added.

See if it fixes for you?

**Comment by tpapp** [comment]
@vchuravy, @wsmoses, thank you very much for looking into this. I still have the issue, ie
```julia
[2174897] signal (11.1): Segmentation fault
in expression starting at /home/tamas/code/julia/misc/EnzymeBugMWE.jl/replicate_crash.jl:58
unknown function (ip: 0x7f500000ac28)
unknown function (ip: 0x402fffffffffffff)
unknown function (ip: 0x7f500000a529)
macro expansion at /home/tamas/.julia/packages/Enzyme/6zfqp/src/compiler.jl:9515 [inlined]
enzyme_call at /home/tamas/.julia/packages/Enzyme/6zfqp/src/compiler.jl:9207
unknown function (ip: 0x7f4fcc940cda)
jl_apply at /home/tamas/src/julia-1.9/src/julia.h:1879 [inlined]
do_apply at /home/tamas/src/julia-1.9/src/builtins.c:730
AdjointThunk at /home/tamas/.julia/packages/Enzyme/6zfqp/src/compiler.jl:9176
top-level scope at /home/tamas/code/julia/misc/EnzymeBugMWE.jl/replicate_crash.jl:63
jl_toplevel_eval_flex at /home/tamas/src/julia-1.9/src/toplevel.c:903
jl_toplevel_eval_flex at /home/tamas/src/julia-1.9/src/toplevel.c:856
ijl_toplevel_eval_in at /home/tamas/src/julia-1.9/src/toplevel.c:971
eval at ./boot.jl:370 [inlined]
include_string at ./loading.jl:1899
_include at ./loading.jl:1959
include at ./Base.jl:457
jfptr_include_40365 at /home/tamas/src/julia-1.9/usr/lib/julia/sys.so (unknown line)
exec_options at ./client.jl:307
_start at ./client.jl:522
```
using the replication code in the `tp/gcdbg2` branch of the repo above. But now it takes more runs to get a crash (about 10-20).

I am using Enzyme 78f151afaf4690ac20450ef50d9d24c9cf614aa3 and Enzyme_jll 3f279551013144a202d900222c32e5e06c30a0b6, basically from updating Enzyme#main this morning. Is this the right version, or should I wait until it trickles through?

**Comment by tpapp** [comment]
I just tried Enzyme#master and I still have the issue. Updated the `tp/gdcbg2` branch in the private repo above.

I understand that this is something outside Enzyme, but can you please link the relevant LLVM issue (if you opened one) so that I can follow how this develops?

**Comment by wsmoses** [comment]
After several weeks worth of deep GC debugging, we have a significant fix landed on enzyme proper whose jll will hopefully hit here in a day or two.

Not sure if it fixes this issue specifically, but I we definitely should retry once it lands here.

In particular it is aimed to fix that GC issue which was messing with our CI and blocking us from cutting a new release.

**Comment by wsmoses** [comment]
Okay the other GC fix has landed, @tpapp can you check if it fixes your/this issue as well?

**Comment by tpapp** [comment]
yes, the issue is fixed now. thank you very much for all the work.
